lsp-types/src/Language/LSP/Types/Uri.hs

```diff
@@ -151,37 +184,78 @@ platformAdjustToUriPath systemOS srcPath
           FPP.addTrailingPathSeparator (init drv)
       | otherwise = drv
 
+#ifdef OS_PATH
+type OsPath = OsPath.OsPath
+#else
+type OsPath = FilePath
+#endif
+
 -- | Newtype wrapper around FilePath that always has normalized slashes.
 -- The NormalizedUri and hash of the FilePath are cached to avoided
 -- repeated normalisation when we need to compute them (which is a lot).
 --
 -- This is one of the most performance critical parts of ghcide, do not
 -- modify it without profiling.
-data NormalizedFilePath = NormalizedFilePath NormalizedUri !FilePath
+data NormalizedFilePath = NormalizedFilePath !NormalizedUri {-# UNPACK #-} !ShortByteString
```

Comment:
How much of a difference does this make over just doing `!OsPath`? The unpack is nice, I guess the annoyance is that if we leave it in we'll get a warning in the case where it's `FilePath`? Or is `OsPath` not unpackable also? Just trying to keep this simple, if we don't do the unpacking then we save on some more CPP...

Reply:
The point here is that I always want to use `ShortByteString` as the underlying type, no matter if the exposed type is `OsPath` or `FilePath`. `{-# UNPACK #-}` is only an optimization though. Then all versions of GHC benefit from the same performance gain, in terms of memory usage.

Reply:
Sorry, I'm still slightly confused. Which of these is it?
1. Using `OsPath` provides no memory usage benefits even with new `filepath`, so we need to use `ShortByteString` regardless to get benefits.
2. Using `OsPath` does provide memory usages in itself, but if we use `ShortByteString` we can get the same improvements for old GHCs as well.

I do like this a lot better than the `filepath-compat` version, so I think I'm probably okay with doing this even if it's only for 2.

Reply:
It's 2. Another option would be conditionally using `ShortByteString`/`OsPath` in `NormalizedFilePath`. But I think it will introduce more CPP. A stable type is easier to work with.

Reply:
Or just only getting the performance improvements in later GHCs. But the changes here seem localized enough that it's probably worth doing.